### PR TITLE
Fix #2125: clj-kondo.hooks-api/*reload* doesn't work as expected

### LINKED ISSUE #2125: Setting clj-kondo.hooks-api/*reload* to true does not lint with the latest hook changes.
**version**

clj-kondo v2023.05.26

**platform**

JVM

**problem**

Binding `clj-kondo.hooks-api/*reload*` to `true` isn't using the new evaluated hook code on REPL.

**repro**

Follow the instructions on [Developing hooks in the REPL](https://github.com/clj-kondo/clj-kondo/blob/master/doc/hooks.md#developing-hooks-in-the-repl)

I created the following helper on a test namespace:
```
(defn ^:private findings [code]
  (:findings (binding [clj-kondo.hooks-api/*reload* true]
               (with-in-str
                 (pr-str code)
                 (clj-kondo.core/run! {:lint ["-"]
                                       :cache false
                                       :config-dir (io/file "resources/clj-kondo.exports/my-lib")})))))

```

**expected behavior**

While developing a hook, I decided to write a few unit tests for it. In order to have a short feedback cycle, I was running the tests on the REPL and set `clj-kondo.hooks-api/*reload*` to true. I was expecting that after evaluating changes on the new hook when I run the tests it will use the newer definition but it isn't.

**suggestion**

I read the code and found out [this binding](https://github.com/clj-kondo/clj-kondo/blob/a7ee15500a58aab8142eaf324b65aa657e5332e9/src/clj_kondo/impl/hooks.clj#L157-L159) is being resolved on compile time to the memoized function. So it isn't being reevaluated on runtime.

I've cooked a fix for that https://github.com/clj-kondo/clj-kondo/pull/2123

### PR DESCRIPTION
**Pull request by acamargo**
Please answer the following questions and leave the below in as part of your PR.

- [x] I have read the [developer documentation](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md).

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).
https://github.com/clj-kondo/clj-kondo/issues/2125

- [ ] This PR contains a [test](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#tests) to prevent against future regressions

- [x] I have updated the [CHANGELOG.md](https://github.com/clj-kondo/clj-kondo/blob/master/CHANGELOG.md) file with a description of the addressed issue.

### CODE REVIEW

**Comment by borkdude** [comment]
LGTM

**Comment by acamargo** [comment]
Hi @borkdude , I was wondering, how can I test this change?

**Comment by borkdude** [comment]
I'm fine with merging this if you tested it locally.

**Review by ericdallo — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ For a list of breaking changes, check [here](#breaking-changes).
 - [#256](https://github.com/clj-kondo/clj-kondo/issues/256): warn on reader conditional usage in non-cljc files
 - [#2115](https://github.com/clj-kondo/clj-kondo/issues/2115): false positive `:redundant-fn-wrapper` in CLJS when passing keyword to JS
 - [#1082](https://github.com/clj-kondo/clj-kondo/issues/1082): protocol methods do not support varargs
+- [#2125](https://github.com/clj-kondo/clj-kondo/issues/2125): clj-kondo.hooks-api/*reload* doesn't work as expected
```
Comment (borkdude) [inline CHANGELOG.md:33]: Can you perhaps be more specific here about what didn't work as expected?

Comment (acamargo) [reply CHANGELOG.md:33]: I asked ChatGPT for help and it suggested: "Setting `clj-kondo.hooks-api/*reload*` to true does not reflect the latest hook changes." :-)

Comment (acamargo) [reply CHANGELOG.md:33]: Perhaps "Setting `clj-kondo.hooks-api/*reload*` to true does not lint with latest hook changes."

Comment (acamargo) [inline CHANGELOG.md:33]:
```suggestion
- [#2125](https://github.com/clj-kondo/clj-kondo/issues/2125): Setting `clj-kondo.hooks-api/*reload*` to true does not lint with the latest hook changes.
```

**Pull request merged**